File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/AsStringDeserializer.scala
package com.exasol.cloudetl.kafka.deserialization

import org.apache.kafka.common.serialization.Deserializer
import org.apache.kafka.common.serialization.StringDeserializer

/**
 * Emits the string representation record as single element sequence.
 */
class AsStringDeserializer(fieldSpecs: Seq[FieldSpecification])
    extends Deserializer[Map[FieldSpecification, Seq[Any]]] {

  private[this] val deserializer = new StringDeserializer

  override final def deserialize(
    topic: String,
    data: Array[Byte]
  ): Map[FieldSpecification, Seq[Any]] =
    fieldSpecs.map(fieldSpec => (fieldSpec, Seq(deserializer.deserialize(topic, data)))).toMap
}

File: src/main/java/com/exasol/cloudetl/kafka/deserialization/AsStringDeserializer.java
package com.exasol.cloudetl.kafka.deserialization;

import java.util.*;

import org.apache.kafka.common.serialization.Deserializer;
import org.apache.kafka.common.serialization.StringDeserializer;

public class AsStringDeserializer implements Deserializer<Map<FieldSpecification, List<Object>>> {
    private final List<FieldSpecification> fieldSpecs;
    private final StringDeserializer deserializer = new StringDeserializer();

    public AsStringDeserializer(final List<FieldSpecification> fieldSpecs) {
        this.fieldSpecs = fieldSpecs;
    }

    @Override
    public Map<FieldSpecification, List<Object>> deserialize(final String topic, final byte[] data) {
        final Map<FieldSpecification, List<Object>> result = new LinkedHashMap<>();
        final Object value = this.deserializer.deserialize(topic, data);
        for (final FieldSpecification fieldSpec : this.fieldSpecs) {
            result.put(fieldSpec, Collections.singletonList(value));
        }
        return result;
    }
}
